src/libraries/Common/src/Interop/Unix/System.Native/Interop.GNameUName.cs

```diff
@@ -0,0 +1,73 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.InteropServices;
+using System.Buffers;
+using System.Text;
+using System;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        /// <summary>
+        /// Gets the group name associated to the specified group ID.
+        /// </summary>
+        /// <param name="gid">The group ID.</param>
+        /// <returns>On success, return a string with the group name. On failure, returns a null string.</returns>
+        internal static string? GetGName(uint gid) => GetGNameOrUName(gid, isGName: true);
+
+        /// <summary>
+        /// Gets the user name associated to the specified user ID.
+        /// </summary>
+        /// <param name="uid">The user ID.</param>
+        /// <returns>On success, return a string with the user name. On failure, returns a null string.</returns>
+        internal static string? GetUName(uint uid) => GetGNameOrUName(uid, isGName: false);
+
+        private static string? GetGNameOrUName(uint id, bool isGName)
+        {
+            // Common max name length, like /etc/passwd, useradd, groupadd
+            int outputBufferSize = 32;
+
+            while (true)
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(outputBufferSize);
+                try
+                {
+                    int resultLength = isGName ?
+                        Interop.Sys.GetGName(id, buffer, buffer.Length) :
+                        Interop.Sys.GetUName(id, buffer, buffer.Length);
+
+                    if (resultLength < 0)
+                    {
+                        // error
+                        return null;
+                    }
+                    else if (resultLength < buffer.Length)
+                    {
+                        // success
+                        return Encoding.UTF8.GetString(buffer, 0, resultLength);
+                    }
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
+                }
+
+                // Output buffer was too small, loop around again and try with a larger buffer.
+                outputBufferSize = buffer.Length * 2;
+
+                if (outputBufferSize > 256) // Upper limit allowed for login name in kernel
```

Comment:
If we are limiting to 256 anyway, can't we just `stackalloc byte[256]` and use that? Why go through the pain of a loop and ArrayPool Rent/Return?

Reply:
Unless it's explicitly documented (pretty much, there's some define -- eg NI_MAXHOST for getnameinfo()) I recommend that we use the stackalloc then resize loop, as in this example:

https://github.com/dotnet/runtime/blob/43dd0a74ab524278620d8c6a9d33a9b73b2d2228/src/libraries/Common/src/Interop/Unix/System.Native/Interop.GetCwd.cs#L18-L45

I did find at least one case in the pal where it mallocs the buffer and resizes it for you, but it seems our pattern is to do all this in managed code.

Ideally it returns the actual length it needs, or you only loop on ERANGE. I'm not sure what is the pattern.

Reply:
We use a similar loop in `ReadLink`:

https://github.com/dotnet/runtime/blob/c2ec86b1c552ac8a1749f9f98e012f707e325660/src/libraries/Common/src/Interop/Unix/System.Native/Interop.ReadLink.cs#L31-L68

Reply:
Unlike the `GetCwd` and the `ReadLink` examples, which do not have an upper bound on the length of the buffer, the uname and gname case do have a max length. So I am now more inclined to just use the 256 byte limit.

Reply:
Where is that length documented?

Reply:
https://serverfault.com/questions/294121/what-is-the-maximum-username-length-on-current-gnu-linux-systems 

The above page says that:

- `man useradd` indicates a max length of 32. Confirmed in the `useradd` manual https://linux.die.net/man/8/useradd  and the `groupadd` manual: https://linux.die.net/man/8/groupadd
- The centos kernel max login length is 256, indicated in the `local_lim.h` file as `LOGIN_NAME_MAX`. I found this version of that file that confirms this: https://android.googlesource.com/platform/prebuilts/gcc/linux-x86/host/x86_64-linux-glibc2.7-4.6/+/02075080d51c371ae87b9898bf84a085e436ee27/sysroot/usr/include/bits/local_lim.h#83
- The `/etc/passwd` file takes usernames with a max length of 32. This illumOS man page confirms this: https://illumos.org/man/5/passwd

Reply:
The limit is implementation specific, so there should be a resize loop.
256 is a good value for the stackalloc to start with because that is the Linux value.

Reply:
Note that the Linux kernel only cares about the numeric ids. libc handles the names.

https://github.com/bminor/glibc/blob/master/sysdeps/unix/sysv/linux/bits/local_lim.h#L89-L90 defines this (implementation specific) limit for Linux:

```c
/* Maximum login name length.  This is arbitrary.  */
#define LOGIN_NAME_MAX		256
```

Reply:
There's another upper limit. It's in the code example in the manual: https://pubs.opengroup.org/onlinepubs/9699919799/functions/getgrgid.html

```cpp
long int initlen = sysconf(_SC_GETGR_R_SIZE_MAX);
size_t len;
if (initlen == -1)
    /* Default initial length. */
    len = 1024;
else
    len = (size_t) initlen;
```

I can use 1024 as the absolute upper limit that would stop the loop.

Reply:
> I can use 1024 as the absolute upper limit that would stop the loop, I think that's large enough (that's the value I got when printing _SC_GETGR_R_SIZE_MAX in Ubuntu).

I don't think that's necessarily the upper limit, eg., [this man page for getgrnam_r](https://www.man7.org/linux/man-pages/man3/getgrnam.3.html) says 

> The call
           sysconf(_SC_GETGR_R_SIZE_MAX)
       returns either -1, without changing [errno](https://www.man7.org/linux/man-pages/man3/errno.3.html), or an initial
       suggested size for buf.  (If this size is too small, the call
       fails with ERANGE, in which case the caller can retry with a
       larger buffer.)

In the [man page for getgrgid](https://www.man7.org/linux/man-pages/man3/getgrgid.3p.html) it says it in different words, again referring to getgrnam_r

> A call to sysconf(_SC_GETGR_R_SIZE_MAX) returns either -1
       without changing [errno](https://www.man7.org/linux/man-pages/man3/errno.3.html) or an initial value suggested for the size
       of this buffer.

Although I see Mono just used it (or 1024) and[ did not resize](https://github.com/mono/mono/blob/5d2e3bc3b3c8184d35b2f7801e88d96470d367c4/mono/metadata/mono-security.c#L449) without provision for resize, here is a bug report where it was not large enough. https://bugs.ruby-lang.org/issues/9600